src/game.zig

```diff
@@ -277,6 +278,66 @@ pub const collision = struct {
 		return resultBox;
 	}
 
+	pub fn calculateFriction(comptime side: main.utils.Side, pos: Vec3d, hitBox: Box, defaultFriction: f32) f32 {
+		const boundingBox: Box = .{
+			.min = pos + hitBox.min,
+			.max = pos + hitBox.max,
+		};
+		const minX: i32 = @intFromFloat(@floor(boundingBox.min[0]));
+		const maxX: i32 = @intFromFloat(@floor(boundingBox.max[0] - 0.0001));
+		const minY: i32 = @intFromFloat(@floor(boundingBox.min[1]));
+		const maxY: i32 = @intFromFloat(@floor(boundingBox.max[1] - 0.0001));
+
+		const z: i32 = @intFromFloat(@floor(boundingBox.min[2] - 0.01));
+
+		var friction: f32 = 0;
+		var totalArea: f64 = 0;
+
+		{
+			var x: i32 = minX;
+			while (x <= maxX) : (x += 1) {
+				var y: i32 = minY;
+				while (y <= maxY) : (y += 1) {
+					const _block = if(side == .client) main.renderer.mesh_storage.getBlock(x, y, z)
+					else main.server.world.?.getBlock(x, y, z);
+					const min = @max(Vec2d{@floatFromInt(x), @floatFromInt(y)}, vec.xy(boundingBox.min));
+					const max = @min(Vec2d{@floatFromInt(x + 1), @floatFromInt(y + 1)}, vec.xy(boundingBox.max));
+					const area = (max[0] - min[0]) * (max[1] - min[1]);
+
+					if (_block) |block| {
+						if (block.collide()) {
+							totalArea += area;
```

Comment:
Please merge this with the loop below.
Dividing friction by totalArea after the loop should be enough to do this.